README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# fdakma

<!-- badges: start -->

[![R-CMD-check](https://github.com/astamm/fdakma/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/astamm/fdakma/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

The **fdakma** package is only a wrapper around the faster and more
complete
[**fdacluster**](https://astamm.github.io/fdacluster/index.html)
package, which exposes the k-means alignment algorithm to the user. It
is mainly kept alive because it was advertised under this name when the
seminal paper was published. We strongly encourage R users to switch to
using the
[**fdacluster**](https://astamm.github.io/fdacluster/index.html) from
now on.

## Installation

You can install the development version of **fdakma** from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("astamm/fdakma")
```

## Example

``` r
library(fdakma)
#> ! The fdakma package is only a wrapper around the
#> faster and more complete fdacluster package, which exposes the k-means
#> alignment algorithm to the user. It is mainly kept alive because it was
#> advertised under this name when the seminal paper was published. We strongly
#> encourage R users to switch to using the fdacluster from now on.

res <- kma(
  fdacluster::simulated30$x,
  fdacluster::simulated30$y,
  seeds = c(1, 21),
  n_clust = 2,
  center_method = "medoid",
  warping_method = "affine",
  dissimilarity_method = "pearson"
)
#> Information about the data set:
#>  - Number of observations: 30
#>  - Number of dimensions: 1
#>  - Number of points: 200
#> 
#> Information about cluster initialization:
#>  - Number of clusters: 2
#>  - Initial seeds for cluster centers:          1        21
#> 
#> Information about the methods used within the algorithm:
#>  - Warping method: affine
#>  - Center method: medoid
#>  - Dissimilarity method: pearson
#>  - Optimization method: bobyqa
#> 
#> Information about warping parameter bounds:
#>  - Warping options:    0.1500   0.1500
#> 
#> Information about convergence criteria:
#>  - Maximum number of iterations: 100
#>  - Distance relative tolerance: 0.001
#> 
#> Information about parallelization setup:
#>  - Number of threads: 1
#>  - Parallel method: 0
#> 
#> Other information:
#>  - Use fence to robustify: 0
#>  - Check total dissimilarity: 1
#>  - Compute overall center: 0
#> 
#> Running k-centroid algorithm:
#>  - Iteration #1
#>    * Size of cluster #0: 20
#>    * Size of cluster #1: 10
#>  - Iteration #2
#>    * Size of cluster #0: 20
#>    * Size of cluster #1: 10
#> 
#> Active stopping criteria:
#>  - Memberships did not change.
```

``` r
plot(res, type = "data")
```

<img src="man/figures/README-unnamed-chunk-3-1.png" width="100%" />

``` r
plot(res, type = "warping")
```

<img src="man/figures/README-unnamed-chunk-4-1.png" width="100%" />


NEWS.md
# fdakma 1.3.1

* Fix CRAN issues.

# fdakma 1.3.0

* Added a `NEWS.md` file to track changes to the package.
* Rely on [**fdacluster**](https://astamm.github.io/fdacluster/) package. Please use this package from now on.


R/kma.compare.R
kma.compare <-
function (x, y0 = NULL, y1 = NULL, n.clust = c(1, 2), warping.method = c("NOalignment", 
    "shift", "dilation", "affine"), similarity.method = "d1.pearson", 
    center.method = "k-means", seeds = NULL, optim.method = "L-BFGS-B", 
    span = 0.15, t.max = 0.1, m.max = 0.1, n.out = NULL, tol = 0.01, 
    fence = TRUE, iter.max = 100, show.iter = 0, plot.graph = 0, 
    nstart = 1, return.all = FALSE) 
{
    if (length(n.clust) == 0) {
        stop("length of n.clust must be positive")
    }
    for (ind in 1:length(n.clust)) {
        if (n.clust[ind] <= 0) {
            stop("n.clust must be positive")
        }
        if (n.clust[ind] != floor(n.clust[ind]) | n.clust[ind] != 
            ceiling(n.clust[ind])) {
            warning("the cluster number must be integer, the value has been approximated to the nearest integer to continue")
            n.clust[ind] <- round(n.clust[ind])
        }
    }
    if (return.all == TRUE) {
        warning("return.all assigned to FALSE, only the best result in each case will be considered")
    }
    n.clust <- sort(n.clust)
    sim <- NULL
    sim_NOalignment <- NULL
    sim_ONLYSHI <- NULL
    sim_ONLYDIL <- NULL
    Result.NOalignment <- NULL
    Result.shift <- NULL
    Result.dilation <- NULL
    Result.affine <- NULL
    if (!"NOalignment" %in% warping.method && !"shift" %in% warping.method && 
        !"dilation" %in% warping.method && !"affine" %in% warping.method) {
        stop("warping.method does not contain any accepted values")
    }
    cycle <- NULL
    colors <- NULL
    if ("NOalignment" %in% warping.method) {
        cycle <- c(cycle, "NOalignment")
        colors <- c(colors, "black")
    }
    if ("shift" %in% warping.method) {
        cycle <- c(cycle, "shift")
        colors <- c(colors, "blue")
    }
    if ("dilation" %in% warping.method) {
        cycle <- c(cycle, "dilation")
        colors <- c(colors, "forestgreen")
    }
    if ("affine" %in% warping.method) {
        cycle <- c(cycle, "affine")
        colors <- c(colors, "orange")
    }
    if (length(dim(y0)) != 0) {
        n.obs <- dim(y0)[1]
    }
    if (length(dim(y1)) != 0) {
        n.obs <- dim(y1)[1]
    }
    if (is.null(seeds)) {
        seeds <- sample(1:n.obs, max(n.clust))
    }
    if (length(dim(seeds)) == 0) {
        seeds <- as.matrix(t(seeds))
    }
    if (length(seeds[1, ]) > max(n.clust)) {
        stop("number of columns of \"seeds\" must be inferior or equal to max(n.clust)")
    }
    if (dim(seeds)[2] > nstart) {
        warning("Number of row of seeds higher than nstart, only the first nstart rows of seeds will be considered")
        seeds <- seeds[1:nstart, ]
        if (length(dim(seeds)) == 0) {
            seeds <- as.matrix(t(seeds))
        }
    }
    if (length(seeds[1, ]) < max(n.clust)) {
        seeds_2 <- matrix(NA, nrow = nstart, ncol = max(n.clust))
        seeds_2[1:dim(seeds)[1], 1:dim(seeds)[2]] <- seeds
        seeds_2[which(is.na(seeds_2))] <- sample(1:n.obs, length(which(is.na(seeds_2))))
        seeds <- seeds_2
    }
    if (length(seeds) != 0) {
        for (contr in 1:length(seeds)) {
            if (seeds[contr] > n.obs || seeds[contr] <= 0) {
                stop("At least a value of \"seeds\" is not valid (is negative or null or superior to the number of observations)")
            }
        }
    }
    optim.method.available <- c("L-BFGS-B", "SANN")
    if (length(optim.method) != 0) {
        if (!optim.method %in% optim.method.available) {
            stop("Value of \"optim.method\" not valid. If defined, it must be one of the following methods (optim package methods): ", 
                "\"", optim.method.available[1], "\"", " ", "\"", 
                optim.method.available[2], "\"")
        }
    }
    if (length(optim.method) == 0) {
        optim.method <- "L-BFGS-B"
    }
    similarity.method.compare <- similarity.method
    for (k in n.clust) {
        for (warping.method.tra in cycle) {
            return.all <- FALSE
            Result <- kma(x = x, y0 = y0, y1 = y1, n.clust = k, 
                warping.method = warping.method.tra, similarity.method = similarity.method.compare, 
                center.method = center.method, seeds = seeds[, 
                  1:k], optim.method = optim.method, span = span, 
                t.max = t.max, m.max = m.max, n.out = n.out, 
                tol = tol, fence = fence, iter.max = iter.max, 
                show.iter = show.iter, nstart = nstart, return.all = return.all)
            if (warping.method.tra == "NOalignment") {
                sim_NOalignment <- c(sim_NOalignment, mean(Result$similarity.final))
                Result.NOalignment <- c(Result.NOalignment, list(Result))
            }
            if (warping.method.tra == "shift") {
                sim_ONLYSHI <- c(sim_ONLYSHI, mean(Result$similarity.final))
                Result.shift <- c(Result.shift, list(Result))
            }
            if (warping.method.tra == "dilation") {
                sim_ONLYDIL <- c(sim_ONLYDIL, mean(Result$similarity.final))
                Result.dilation <- c(Result.dilation, list(Result))
            }
            if (warping.method.tra == "affine") {
                sim <- c(sim, mean(Result$similarity.final))
                Result.affine <- c(Result.affine, list(Result))
            }
        }
    }
    if (plot.graph == 1 && length(n.clust) > 1) {
        dev.new()
        massimo <- max(sim_NOalignment, sim_ONLYSHI, sim_ONLYDIL, 
            sim)
        minimo <- min(sim_NOalignment, sim_ONLYSHI, sim_ONLYDIL, 
            sim)
        for (indice in 1:length(cycle)) {
            if (indice == 1) {
                if (cycle[indice] == "NOalignment") 
                  plot(sim_NOalignment, type = "b", axes = F, 
                    ylim = c(minimo, massimo), xlab = "", ylab = "", 
                    col = "black", lwd = 4)
                if (cycle[indice] == "shift") 
                  plot(sim_ONLYSHI, type = "b", axes = F, ylim = c(minimo, 
                    massimo), xlab = "", ylab = "", col = "blue", 
                    lwd = 4)
                if (cycle[indice] == "dilation") 
                  plot(sim_ONLYDIL, type = "b", axes = F, ylim = c(minimo, 
                    massimo), xlab = "", ylab = "", col = "forestgreen", 
                    lwd = 4)
                if (cycle[indice] == "affine") 
                  plot(sim, type = "b", axes = F, ylim = c(minimo, 
                    massimo), xlab = "", ylab = "", col = "orange", 
                    lwd = 4)
            }
            else {
                if (cycle[indice] == "NOalignment") 
                  points(sim_NOalignment, type = "b", ylim = c(minimo, 
                    massimo), xlab = "", ylab = "", col = "black", 
                    lwd = 4)
                if (cycle[indice] == "shift") 
                  points(sim_ONLYSHI, type = "b", ylim = c(minimo, 
                    massimo), xlab = "", ylab = "", col = "blue", 
                    lwd = 4)
                if (cycle[indice] == "dilation") 
                  points(sim_ONLYDIL, type = "b", ylim = c(minimo, 
                    massimo), xlab = "", ylab = "", col = "forestgreen", 
                    lwd = 4)
                if (cycle[indice] == "affine") 
                  points(sim, type = "b", ylim = c(minimo, massimo), 
                    xlab = "", ylab = "", col = "orange", lwd = 4)
            }
        }
        etichette <- rep(0, length(n.clust))
        for (i in 1:length(n.clust)) {
            etichette[i] <- paste("k =", n.clust[i])
        }
        box()
        boh <- 11
        axis(2)
        axis(1, at = 1:length(n.clust), labels = etichette, las = 0)
        ba <- c("Mean similarity indexes VS Num.clusters", paste("(center.method: ", 
            center.method, ")"))
        title(ba)
        leggenda <- cycle
        if (cycle[1] == "NOalignment") 
            leggenda[1] <- "without alignment"
        if (similarity.method.compare == "d0.L2" || similarity.method.compare == 
            "d1.L2" || similarity.method.compare == "d0.L2.centered" || 
            similarity.method.compare == "d1.L2.centered") {
            legend("topright", inset = 0.01, legend = leggenda, 
                col = colors, lwd = c(3, 3), cex = 0.6)
        }
        else {
            legend("bottomright", inset = 0.01, legend = leggenda, 
                col = colors, lwd = c(3, 3), cex = 0.6)
        }
    }
    if (plot.graph == 1 && length(n.clust) > 1) {
        toppo <- max(sim_NOalignment[1], sim_ONLYSHI[1], sim_ONLYDIL[1], 
            sim[1], na.rm = TRUE)
        uppo <- min(sim_NOalignment[1], sim_ONLYSHI[1], sim_ONLYDIL[1], 
            sim[1], na.rm = TRUE)
        lines(c(1, 1), c(uppo, toppo), col = "darkgrey", type = "l", 
            lty = "dashed", cex = 3, lwd = 3)
    }
    return(list(Result.NOalignment = Result.NOalignment, Result.shift = Result.shift, 
        Result.dilation = Result.dilation, Result.affine = Result.affine, 
        n.clust = n.clust, mean.similarity.NOalignment = sim_NOalignment, 
        mean.similarity.shift = sim_ONLYSHI, mean.similarity.dilation = sim_ONLYDIL, 
        mean.similarity.affine = sim))
}


R/kma.similarity.R
kma.similarity <-
function (x.f = NULL, y0.f = NULL, y1.f = NULL, x.g = NULL, y0.g = NULL, 
    y1.g = NULL, similarity.method, unif.grid = TRUE) {
    similarity.method.available <- c("d0.pearson", "d1.pearson", 
        "d0.L2", "d1.L2", "d0.L2.centered", "d1.L2.centered")
    if (!similarity.method %in% similarity.method.available) {
        stop("Value of \"similarity.method\" not valid. Possibles choices are: ", 
            "\"", similarity.method.available[1], "\"", " ", 
            "\"", similarity.method.available[2], "\"", " ", 
            "\"", similarity.method.available[3], "\"", " ", 
            "\"", similarity.method.available[4], "\"", " ", 
            "\"", similarity.method.available[5], "\"", " ", 
            "\"", similarity.method.available[6], "\"")
    }
    if (similarity.method == "d0.pearson" || similarity.method == 
        "d0.L2" || similarity.method == "d0.L2.centered") {
        work.with.deriv.sim = 0
    }
    if (similarity.method == "d1.pearson" || similarity.method == 
        "d1.L2" || similarity.method == "d1.L2.centered") {
        work.with.deriv.sim = 1
    }
    if (!unif.grid %in% c(TRUE, FALSE)) {
        stop("Value of parameter \"unif.grid\" not valid. \n         It must be TRUE or FALSE")
    }
    var.temp <- FALSE
    var.temp2 <- FALSE
    MAX_NUMBER <- 10^(150)
    if ((length(x.f) == 0 && length(y0.f) == 0) || (length(x.g) == 
        0 && length(y0.g) == 0)) {
        var.temp <- TRUE
        var.temp2 <- TRUE
        warning("Domains of two curves in kma.similarity function do not overlap. If using kma function of fdakma package, suggested reduction of \"t.max\" and/or \"m.max\" or set \"fence\" equal to TRUE")
        if (similarity.method == "d1.pearson" || similarity.method == 
            "d0.pearson") {
            OUT <- 0
        }
        if (similarity.method == "d0.L2" || similarity.method == 
            "d1.L2" || similarity.method == "d0.L2.centered" || 
            similarity.method == "d1.L2.centered") {
            OUT <- MAX_NUMBER
        }
    }
    if (var.temp == FALSE) {
        if (work.with.deriv.sim == 0) {
            if (length(x.f) == 0 || length(y0.f) == 0 || length(x.g) == 
                0 || length(y0.g) == 0) {
                stop("You have to provide the abscissas (x.f and x.g) and the evaluations of functions (y0.f and y0.g) with the similarity.method chosen (", 
                  similarity.method, ")")
            }
            if (length(y1.f) != 0 || length(y1.g) != 0) {
                warning("You provided the evaluations of function first dervivatives (y1.f and/or y1.g) with the chosen similarity.method (", 
                  similarity.method, ").\n                These values has been ignored.")
            }
        }
        if (work.with.deriv.sim == 1) {
            if (length(x.f) == 0 || length(y1.f) == 0 || length(x.g) == 
                0 || length(y1.g) == 0) {
                stop("You have to provide the abscissa (x.f and x.g) and the evaluations of function first derivatives (y1.f and y1.g) with the chosen similarity.method (", 
                  similarity.method, ")")
            }
            if (length(y0.f) != 0 || length(y0.g) != 0) {
                warning("You provided the evaluations of functions (y0.f and/or y0.g) with the chosen similarity.method (", 
                  similarity.method, ").\n                These values has been ignored.")
            }
        }
        if (work.with.deriv.sim == 0) {
            y.f <- y0.f
            y.g <- y0.g
        }
        if (work.with.deriv.sim == 1) {
            y.f <- y1.f
            y.g <- y1.g
        }
        if (length(dim(y.f)) == 2) {
            if (dim(y.f)[1] == 1) {
                y.f <- t(y.f)
                y.g <- t(y.g)
            }
            r.t <- dim(y.f)[2]
            n.camp.f <- dim(y.f)[1]
            n.camp.g <- dim(y.g)[1]
            function1 <- array(0, dim = c(1, n.camp.f, r.t))
            function1[, , 1:r.t] <- y.f
            function2 <- array(0, dim = c(1, n.camp.g, r.t))
            function2[, , 1:r.t] <- y.g
        }
        else {
            r.t <- 1
            if (class(y.f) == "numeric") {
                n.camp.f <- length(y.f)
            }
            else {
                n.camp.f <- dim(y.f)[1]
            }
            if (class(y.g) == "numeric") {
                n.camp.g <- length(y.g)
            }
            else {
                n.camp.g <- dim(y.g)[1]
            }
            function1 <- array(0, dim = c(1, n.camp.f, r.t))
            function1[, , 1:r.t] <- y.f
            function2 <- array(0, dim = c(1, n.camp.g, r.t))
            function2[, , 1:r.t] <- y.g
        }
        if (unif.grid == TRUE) {
            estr.inf.x <- max(min(x.f), min(x.g))
            estr.sup.x <- min(max(x.f), max(x.g))
            precisione.f <- length(intersect(which(x.f >= estr.inf.x), 
                which(x.f <= estr.sup.x)))
            precisione.g <- length(intersect(which(x.g >= estr.inf.x), 
                which(x.g <= estr.sup.x)))
            n.out.sim <- max(precisione.f, precisione.g)
            x.com.sim <- seq(estr.inf.x, estr.sup.x, length = n.out.sim)
            if (length(x.com.sim) <= 1) {
                var.temp <- TRUE
                var.temp2 <- TRUE
                warning("Domains of two curves in kma.similarity function do not overlap. If using kma function of fdakma package, suggested reduction of \"t.max\" and/or \"m.max\" or set \"fence\" equal to TRUE")
                if (similarity.method == "d1.pearson" || similarity.method == 
                  "d0.pearson") {
                  OUT <- 0
                }
                if (similarity.method == "d0.L2" || similarity.method == 
                  "d1.L2" || similarity.method == "d0.L2.centered" || 
                  similarity.method == "d1.L2.centered") {
                  OUT <- MAX_NUMBER
                }
            }
            if (var.temp2 == FALSE) {
                y.f.appr <- array(NA, dim = c(1, length(x.com.sim), 
                  r.t))
                y.g.appr <- array(NA, dim = c(1, length(x.com.sim), 
                  r.t))
                for (l in 1:r.t) {
                  y.f.appr[1, , l] <- approx(x.f, function1[1, 
                    , l], xout = x.com.sim)$y
                  y.g.appr[1, , l] <- approx(x.g, function2[1, 
                    , l], xout = x.com.sim)$y
                }
            }
        }
        else {
            x.com.sim <- x.f
            if (length(x.com.sim) <= 1) {
                warning("Domains of two curves in kma.similarity function do not overlap. If using kma function of fdakma package, suggested reduction of \"t.max\" and/or \"m.max\" or set \"fence\" equal to TRUE")
                if (similarity.method == "d1.pearson" || similarity.method == 
                  "d0.pearson") {
                  OUT <- 0
                }
                if (similarity.method == "d0.L2" || similarity.method == 
                  "d1.L2" || similarity.method == "d0.L2.centered" || 
                  similarity.method == "d1.L2.centered") {
                  OUT <- MAX_NUMBER
                }
            }
            if (var.temp2 == FALSE) {
                y.f.appr <- array(NA, dim = c(1, length(x.com.sim), 
                  r.t))
                y.g.appr <- array(NA, dim = c(1, length(x.com.sim), 
                  r.t))
                for (l in 1:r.t) {
                  y.f.appr[1, , l] <- approx(x.f, function1[1, 
                    , l], xout = x.com.sim)$y
                  y.g.appr[1, , l] <- approx(x.g, function2[1, 
                    , l], xout = x.com.sim)$y
                }
            }
        }
        if (var.temp2 == FALSE) {
            if (similarity.method == "d1.pearson" || similarity.method == 
                "d0.pearson") {
                distance <- 0
                for (l in 1:r.t) {
                  distance <- distance + sum(y.f.appr[1, 
                    , l] * y.g.appr[1, , l])/(sqrt(sum(y.f.appr[1, 
                    , l] * y.f.appr[1, , l])) * sqrt(sum(y.g.appr[1, 
                    , l] * y.g.appr[1, , l])))
                }
                OUT <- distance/r.t
            }
            if (similarity.method == "d0.L2" || similarity.method == 
                "d1.L2" || similarity.method == "d0.L2.centered" || 
                similarity.method == "d1.L2.centered") {
                if (similarity.method == "d0.L2.centered" || 
                  similarity.method == "d1.L2.centered") {
                  for (l in 1:r.t) {
                    y.f.appr[1, , l] <- y.f.appr[1, , l] - mean(y.f.appr[1, 
                      , l], na.rm = TRUE)
                    y.g.appr[1, , l] <- y.g.appr[1, , l] - mean(y.g.appr[1, 
                      , l], na.rm = TRUE)
                  }
                }
                D <- abs(max(x.com.sim) - min(x.com.sim))
                delta <- x.com.sim[2] - x.com.sim[1]
                L2distance <- 0
                for (l in 1:r.t) {
                  L2distance <- L2distance + sqrt(sum(((y.f.appr[1, 
                    2:length(y.f.appr[1, , l]), l] - y.g.appr[1, 
                    2:length(y.f.appr[1, , l]), l])^2) * delta, 
                    na.rm = TRUE))/D
                }
                OUT <- L2distance/r.t
            }
        }
    }
    OUT
}


R/fdakma-internal.R
.Random.seed <-
c(403L, 10L, -368476779L, -1516356180L, 1975654274L, 943062365L, 
-1221381033L, 1605522238L, -731219080L, 1056585819L, -1166901319L, 
-194232824L, -1488950090L, -622101519L, -2147138013L, -575932814L, 
202190020L, 765766279L, -143023523L, 19866708L, -2126303622L, 
-45806139L, 236302111L, -378103130L, -868279248L, -16285997L, 
-896344847L, -2033698784L, 785424062L, 742947529L, -293662597L, 
2112802378L, 1207040684L, 950414767L, -208986779L, -1852388580L, 
-795791502L, 757515245L, 1819280423L, 263248974L, -545218392L, 
1569263275L, 117926857L, 2032469976L, 1887784582L, 784242529L, 
1202883315L, 542578018L, 2107892692L, -1637349673L, -360747827L, 
-141530012L, -352443830L, -678451275L, -909346033L, -438679530L, 
668340928L, 2081956995L, 1769296033L, 458856080L, 1775306094L, 
622916377L, -444631125L, -1081141862L, 990277980L, 1532790879L, 
236521525L, -1971303476L, -367231198L, 1853350269L, -955479433L, 
1945621406L, -825204136L, 1728039867L, -573145767L, -1709148440L, 
-1284094186L, 1640499345L, 688391491L, 1641201746L, -1676860700L, 
-677022425L, 562482557L, 1196082292L, -1378480550L, -1613006171L, 
1653850303L, -896515258L, 518885840L, 435368947L, -555087087L, 
1636367168L, -2024471778L, 968903657L, -1531641573L, 1105640234L, 
-1952285620L, 342504463L, -1587265723L, -447833860L, -1655593134L, 
-1733203763L, -1406257017L, -947147922L, -968966840L, 1931848011L, 
-1024394071L, 1761680888L, 257115430L, -401459647L, -143376429L, 
-2142169726L, -432244300L, -306608841L, -1128083411L, -1458168508L, 
-1135830422L, -366119723L, -139820433L, 1972479862L, -1101805408L, 
-1121558173L, 491991553L, -159563408L, -452992754L, -668363911L, 
-2140349685L, 634340410L, 801792828L, -1091749761L, -32138283L, 
294237676L, -146819262L, 1089665181L, 1197670679L, -156264322L, 
25686072L, -1175771365L, 362390649L, -451629880L, -274432778L, 
165199153L, 1524850019L, -868712654L, 550385796L, -1376557881L, 
-432404579L, -1332341996L, -2104065606L, -14346235L, 186658271L, 
2103025766L, -644018192L, 105310483L, -1515137359L, -204889760L, 
333338366L, -1847348599L, -834007621L, 562403210L, -1232560532L, 
922615791L, 1829574693L, 1786607836L, 1997184690L, 836948397L, 
1085347175L, 293724814L, 1641458536L, 2099106667L, -246141687L, 
-1847319656L, -505859386L, 1987541025L, -1336390477L, 300535330L, 
-1161834476L, 728323223L, -116457843L, 1884237732L, 970437002L, 
353103349L, -609025969L, -1708099754L, -851513856L, 352591683L, 
-92391711L, -383560624L, 777002158L, 1555359833L, -1885566869L, 
562499034L, -1109248356L, 1634338975L, 385106677L, -1574518900L, 
72281698L, -673639875L, 1478286519L, 802333534L, 786384536L, 
1687184891L, -1644529767L, 1220968488L, 961743062L, 1273924945L, 
-433497085L, -1985739886L, -1013583580L, 319449063L, -1001202627L, 
-205419852L, 1518840090L, -946951579L, 808836863L, 888677254L, 
598680848L, 725024179L, 629595729L, 2053037568L, -510461218L, 
653728041L, -2053856037L, -695344150L, 753980300L, 891819215L, 
-437486459L, 433944380L, -2130450926L, -292012614L, 1951876560L, 
1455648748L, -1299952748L, -2118340398L, -107594656L, -1449102004L, 
-330790112L, -1548043646L, 811622888L, -887544852L, 1476777532L, 
-472580494L, 155120496L, 1399409956L, -359212488L, -2042048102L, 
-2090449456L, -34413124L, 1228048020L, 1851966082L, 229545408L, 
-710906052L, 916288144L, -1088182494L, 570185224L, 1675100940L, 
-1998939236L, -1841380334L, -938536064L, 964299540L, -1029671640L, 
-1902243782L, 1081537488L, -1449108884L, -1333445580L, 323491986L, 
-1063173792L, -1192219316L, -1631875616L, 801715106L, -536310872L, 
1309887628L, -1980628036L, 1909849202L, -251602960L, 1110053956L, 
1576876120L, -188030918L, 840544368L, -133441988L, 777246484L, 
-1854969150L, 1755009760L, 18664572L, -647004976L, 1891470242L, 
-240603608L, 989329676L, -1320986692L, -1506055886L, 389341376L, 
396198452L, -705557816L, 485849786L, 138347664L, -473385236L, 
-1072931244L, -1981462894L, -142944480L, -330594356L, -119267104L, 
-595788094L, 50168872L, -1120449620L, -55174468L, 1429293554L, 
-1750985168L, -1689106524L, 1785041336L, 360051546L, -1588218736L, 
-1764394116L, 1695446804L, 1510870338L, 1950891776L, 400339900L, 
-41971888L, -583608030L, -667704760L, -1147359220L, -1100565412L, 
-1614970414L, -305373696L, -1418481324L, -1310592792L, -1224257286L, 
1447691984L, -465120980L, 519409908L, -727798382L, 747465056L, 
-1039751156L, -1519623456L, -992363934L, -25603416L, 1112877644L, 
1076043644L, -962132686L, 1113405040L, -680733884L, 897984472L, 
-1410050182L, 1392404016L, -1739925572L, 1274279508L, -1579605950L, 
-1868877664L, -1360136516L, 1807088080L, 367678306L, 1757278952L, 
-2107432756L, 1644241404L, -2089410062L, -359306624L, 525540852L, 
315041416L, -1009541318L, -1035988144L, -1239757716L, -1802980972L, 
1392211026L, 1374175712L, 1454828748L, 220398752L, 730793346L, 
1011378408L, -1388947732L, -174521028L, 52567154L, 607074416L, 
1292365860L, 111025208L, -386830566L, 2082876624L, -776143044L, 
1507309460L, 1158718338L, -2044155456L, -624738756L, -1282447984L, 
-2036703710L, 810599688L, 278036108L, 580333340L, -817004398L, 
-1158858368L, -774400492L, -1928679640L, 141371962L, -238584368L, 
1159743852L, -641121356L, 68974610L, -1301428768L, 1034180556L, 
-6700832L, -429213662L, 1463234856L, -313979380L, -1340135620L, 
-858319886L, -1391990032L, -1719046588L, -1968074152L, -202170566L, 
-96692752L, 441169596L, -75048684L, -632057150L, 2045461600L, 
-775810180L, -1057036336L, -1987940062L, 1212594856L, 903084940L, 
-406681412L, -1345929422L, -857118144L, -76784460L, -1644842424L, 
428455354L, 2140340752L, 907895532L, 424239444L, 854040850L, 
-1733038688L, 607580876L, -231113120L, -124299326L, -223561944L, 
671068076L, -1361349828L, -824129166L, 735632176L, -1382646492L, 
-1728984264L, -448057638L, 1825376144L, 469847804L, 1697807252L, 
1773826114L, -1621420800L, 1814660156L, -1810628272L, 1685864610L, 
-1914638008L, -327350516L, -384023844L, -1760928430L, 973575296L, 
694382676L, -163278872L, -2121848582L, -1318462640L, -53969108L, 
1537167811L, 131719173L, -491764830L, -1123741808L, 1716497209L, 
-1678213141L, 535683292L, -1361321718L, 1277702143L, 43745865L, 
1408559070L, 2023352204L, 1201010525L, 89485159L, -816720752L, 
-377357650L, 1501747035L, -2144360643L, 1634178474L, 198548040L, 
-619758991L, 878648291L, 878790372L, 2040602130L, 390883815L, 
782075313L, -1507900426L, -654299628L, 1473589541L, -1166321105L, 
-1191616536L, -467760954L, 542297459L, 1325667797L, 706582962L, 
-1113548928L, -1764357271L, -290668581L, 875030316L, -1991571142L, 
-772605969L, -709180807L, -2142354162L, -1969312548L, -1702034611L, 
-549582153L, -839358400L, 1803913502L, 1563349995L, 1090690925L, 
474306010L, 995731352L, 1509484033L, 1177962419L, -669632812L, 
-210087966L, -1871662473L, -444175423L, -325772986L, -1081292668L, 
967020085L, 986723871L, -123907720L, -2136231850L, -468730653L, 
1032534693L, 477333314L, -590432976L, 159606105L, -1184725301L, 
583974716L, -326379670L, 1102149023L, 327502889L, 64933566L, 
-1470685268L, -1497684739L, -629301177L, 1824186160L, 967968846L, 
-66026373L, -2108020771L, 1050482186L, 1020455720L, -2033019503L, 
749800323L, -714559356L, -1034297678L, 293888519L, -56462767L, 
1660545494L, -340208460L, 676483269L, -591093233L, 125213320L, 
-1239007322L, 1104811859L, 890546357L, 761243154L, 330339296L, 
1776903753L, 47469691L, 2040878412L, -422469414L, 2044358863L, 
-1834873703L, -151097170L, 155594172L, 1139988013L, 449253847L, 
1635276832L, 299498878L, 1803445067L, 1103980877L, 107718522L, 
1354394040L, 265395041L, -2026185965L, 649171764L, 1856909634L, 
-2048662185L, 581632161L, 731720038L, 1416878948L, 1897591445L, 
-692464705L, -1473501352L, 15867254L, 510779267L, -1723980347L, 
209599202L, -1222687920L, -516514183L, 1903460779L, -601188580L, 
1252450378L, -985299905L, 1210999945L, 1149710494L, -1090765492L, 
-1502524131L, 1167394983L, -211531824L, 27488110L, 1872717339L, 
901222269L, 2049480042L, 1089129480L, -1928923983L, -1453616349L, 
1475158052L, -1154551086L, 1791268647L, 1679485297L, -105246794L, 
1466843988L, 1506176485L, -814663697L, 1643976488L, -742084730L, 
-1329005645L, -690377963L, -411695758L, -1016137280L, -903284183L, 
-289202533L, -1878333588L, -387080326L, 1875339439L, -80003267L
)


R/kma.R
#' K-Means Alignment Algorithm
#'
#' This is an implementation of the k-means alignment algorithm originally
#' described in Sangalli et al. (2010), with improvements as proposed in Vantini
#' (2012).
#'
#' @inherit fdacluster::fdakmeans
#' @export
#'
#' @examples
#' #----------------------------------
#' # Extracts 15 out of the 30 simulated curves in `simulated30_sub` data set
#' idx <- c(1:5, 11:15, 21:25)
#' x <- fdacluster::simulated30$x[idx, ]
#' y <- fdacluster::simulated30$y[idx, , ]
#'
#' #----------------------------------
#' # Runs a k-means clustering with affine alignment, searching for 2 clusters
#' res <- kma(
#'   x,
#'   y,
#'   n_clusters = 2,
#'   seeds = c(1, 11),
#'   warping_class = "affine",
#'   centroid_type = "medoid",
#'   metric = "pearson"
#' )
kma <- function(x, y = NULL,
                n_clusters = 1L,
                seeds = NULL,
                seeding_strategy = c("kmeans++", "exhaustive-kmeans++", "exhaustive", "hclust"),
                warping_class = c("affine", "dilation", "none", "shift", "srsf"),
                centroid_type = "mean",
                metric = c("l2", "pearson"),
                cluster_on_phase = FALSE,
                use_verbose = TRUE,
                warping_options = c(0.15, 0.15),
                maximum_number_of_iterations = 100L,
                number_of_threads = 1L,
                parallel_method = 0L,
                distance_relative_tolerance = 0.001,
                use_fence = FALSE,
                check_total_dissimilarity = TRUE,
                compute_overall_center = FALSE,
                add_silhouettes = TRUE,
                expand_domain = TRUE) {
  fdacluster::fdakmeans(
    x = x, y = y,
    n_clusters = n_clusters,
    seeds = seeds,
    seeding_strategy = seeding_strategy,
    warping_class = warping_class,
    centroid_type = centroid_type,
    metric = metric,
    cluster_on_phase = cluster_on_phase,
    use_verbose = use_verbose,
    warping_options = warping_options,
    maximum_number_of_iterations = maximum_number_of_iterations,
    number_of_threads = number_of_threads,
    parallel_method = parallel_method,
    distance_relative_tolerance = distance_relative_tolerance,
    use_fence = use_fence,
    check_total_dissimilarity = check_total_dissimilarity,
    compute_overall_center = compute_overall_center,
    add_silhouettes = add_silhouettes,
    expand_domain = expand_domain
  )
}
